lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php

```diff
@@ -394,15 +451,21 @@ private function addInheritedFields(ClassMetadata $subClass, ClassMetadata $pare
     private function addInheritedRelations(ClassMetadata $subClass, ClassMetadata $parentClass): void
     {
         foreach ($parentClass->associationMappings as $field => $mapping) {
-            if ($parentClass->isMappedSuperclass) {
+            $this->addMappingInheritanceInformation($mapping, $parentClass);
+            // When the class inheriting the relation ($subClass) is the first entity class since the
+            // relation has been defined in a mapped superclass (or in a chain
+            // of mapped superclasses) above, then declare this current entity class as the source of
+            // the relationship.
+            // According to the definitions given in https://github.com/doctrine/orm/pull/10396/,
+            // this is the case <=> ! isset($mapping['inherited']).
+            if (! isset($mapping['inherited'])) {
                 if ($mapping['type'] & ClassMetadata::TO_MANY && ! $mapping['isOwningSide']) {
                     throw MappingException::illegalToManyAssociationOnMappedSuperclass($parentClass->name, $field);
                 }
 
                 $mapping['sourceEntity'] = $subClass->name;
```

Comment:
@mpdude this line seems to be causing the error below, but I don't understand why: both `$mapping['sourceEntity']` and `$subClass->name` are reported as `class-string` by `@psalm-trace`. Help appreciated, but first, I'd like to know if I fixed the conflict properly according to you.

Reply:
Ah it looks like removing the call to `addMappingInheritanceInformation` also fixes this. It uses `mixed[]` so I think that might explain a lot.

Reply:
Conflict resolution looks good :+1:

Reply:
https://github.com/doctrine/orm/pull/10397/files#diff-3e426fe47b58d2d4b970b1b7a0cbf5999af950e4f84e48dce4bee5061b57ed31R361 causes the Psalm error IMHO, `mixed[]` makes Psalm assume the type may change (array shape is lost/amended?)

Reply:
I think it might have to do with the `foreach`, because when I copy `$mapping` to a separate variable, then `psalm-trace` shows what I expect.

Reply:
Copying `$mapping` to some random variable then using that random variable fixes the issue.

Reply:
What's funny is that this is the very method that made me lose a lot of time the other day in https://github.com/doctrine/orm/pull/10405/files#r1072806010

Reply:
I wonder if this should be considered a bug: https://psalm.dev/r/a1767b6886
At the position where I put `psalm-trace`, there is no way there's a `test` key in that array.

Reply:
Not a bug apparently, that's just how `psalm-trace` works… it will describe the variable after the assignment

Reply:
Here's a more exact snippet: https://psalm.dev/r/ca012a1298

Reply:
Here is one that is even closer I think: https://psalm.dev/r/84ffd92a4c

Reply:
And something closerer: https://psalm.dev/r/ebf182b2e6

Reply:
I'll open a bug from this and reference it in `psalm.xml`